media/starboard/starboard_cdm.cc

```diff
@@ -389,190 +341,214 @@ void DrmSystem::OnSessionUpdated(int ticket, SbDrmStatus status,
-    const std::vector<SbDrmKeyStatus>& key_statuses) {
+void StarboardCdm::OnSessionKeyStatusChanged(const std::string& session_id,
+                                             CdmKeysInfo keys_info,
+                                             bool has_additional_usable_key) {
   DCHECK(task_runner_->RunsTasksInCurrentSequence());
-
   LOG(INFO) << "Receiving session key status changed notification from drm"
-            << " system (" << wrapped_drm_system_
-            << "), session id: " << session_id
-            << ", number of key ids: " << key_ids.size();
+            << " system (" << sb_drm_ << "), session id: " << session_id
+            << ", number of key ids: " << keys_info.size()
+            << ", has_additional_usable_key?: "
+            << (has_additional_usable_key ? "true" : "false");
 
   // Find the session by ID.
-  IdToSessionMap::iterator session_iterator =
-      id_to_session_map_.find(session_id);
-  if (session_iterator == id_to_session_map_.end()) {
+  auto session_iterator = session_map_.find(session_id);
+  if (session_iterator == session_map_.end()) {
     LOG(ERROR) << "Unknown session id: " << session_id << ".";
     return;
   }
 
-  // As DrmSystem::Session may be released, need to check it before using it.
-  if (session_iterator->second) {
-    session_iterator->second->update_key_statuses_callback().Run(key_ids,
-                                                                 key_statuses);
-  }
+  StarboardCdmSession* session = session_iterator->second.get();
+  session->InvokeOnSessionKeysChange(has_additional_usable_key,
+                                     std::move(keys_info));
 }
 
-void DrmSystem::OnSessionClosed(const std::string& session_id) {
+void StarboardCdm::OnServerCertificateUpdated(
+    int ticket,
+    SbDrmStatus status,
+    const std::string& error_message) {
   DCHECK(task_runner_->RunsTasksInCurrentSequence());
+  LOG(INFO) << "Receiving server certificate updated notification from drm"
+            << " system (" << sb_drm_ << "), ticket: " << ticket
+            << ", status: " << status << ", error message: " << error_message;
 
-  LOG(INFO) << "Receiving session closed notification from drm system ("
-            << wrapped_drm_system_ << "), session id: " << session_id;
-
-  // Find the session by ID.
-  IdToSessionMap::iterator session_iterator =
-      id_to_session_map_.find(session_id);
-  if (session_iterator == id_to_session_map_.end()) {
-    LOG(ERROR) << "Unknown session id: " << session_id << ".";
+  auto iter = ticket_to_server_certificate_updated_map_.find(ticket);
+  if (iter == ticket_to_server_certificate_updated_map_.end()) {
+    LOG(ERROR) << "Unknown ticket: " << ticket << ".";
     return;
   }
 
-  // As DrmSystem::Session may be released, need to check it before using it.
-  if (session_iterator->second) {
-    session_iterator->second->session_closed_callback().Run();
+  if (status == kSbDrmStatusSuccess) {
```

Comment:
We may need to reject the promise when `status` isn't `kSbDrmStatusSuccess`.  It's great to also check similar places to ensure that they are rejected properly.

Reply:
Done. I've made sure all calls to ResolvePromise have a corresponding RejectPromise.